src/main/java/net/snowflake/client/jdbc/cloud/storage/GcmEncryptionProvider.java

```diff
@@ -0,0 +1,245 @@
+
+    byte[] keyBytes = decryptKey(kekBytes, keyIvBytes, encryptedKeyBytes, keyAad);
+    decryptContentFromFile(file, keyBytes, dataIvBytes, dataAad);
+  }
+
+  static InputStream decryptStream(
+      InputStream inputStream,
+      String encryptedKekBase64,
+      String cekIvBase64,
+      String kekIvBase64,
+      RemoteStoreFileEncryptionMaterial encMat,
+      String dataAad,
+      String keyAad)
+      throws InvalidKeyException, BadPaddingException, IllegalBlockSizeException,
+          InvalidAlgorithmParameterException, NoSuchPaddingException, NoSuchAlgorithmException {
+    byte[] encryptedKeyBytes = base64Decoder.decode(encryptedKekBase64);
+    byte[] ivBytes = base64Decoder.decode(cekIvBase64);
+    byte[] kekIvBytes = base64Decoder.decode(kekIvBase64);
+    byte[] dataAadBytes = base64Decoder.decode(dataAad);
+    byte[] keyAadBytes = base64Decoder.decode(keyAad);
+    byte[] kekBytes = base64Decoder.decode(encMat.getQueryStageMasterKey());
+
+    byte[] fileKeyBytes = decryptKey(kekBytes, kekIvBytes, encryptedKeyBytes, keyAadBytes);
+    return decryptContentFromStream(inputStream, ivBytes, fileKeyBytes, dataAadBytes);
+  }
+
+  private static CipherInputStream decryptContentFromStream(
+      InputStream inputStream, byte[] ivBytes, byte[] fileKeyBytes, byte[] aad)
+      throws InvalidKeyException, InvalidAlgorithmParameterException, NoSuchPaddingException,
+          NoSuchAlgorithmException {
+    GCMParameterSpec gcmParameterSpec = new GCMParameterSpec(TAG_LENGTH, ivBytes);
+    SecretKey fileKey = new SecretKeySpec(fileKeyBytes, 0, fileKeyBytes.length, AES);
+    Cipher fileCipher = Cipher.getInstance(FILE_CIPHER);
+    fileCipher.init(Cipher.DECRYPT_MODE, fileKey, gcmParameterSpec);
+    if (aad != null) {
+      fileCipher.updateAAD(aad);
+    }
+    return new CipherInputStream(inputStream, fileCipher);
+  }
+
+  private static void decryptContentFromFile(
+      File file, byte[] fileKeyBytes, byte[] cekIvBytes, byte[] aad)
+      throws InvalidKeyException, InvalidAlgorithmParameterException, IOException,
+          NoSuchPaddingException, NoSuchAlgorithmException {
+    SecretKey fileKey = new SecretKeySpec(fileKeyBytes, AES);
+    GCMParameterSpec gcmParameterSpec = new GCMParameterSpec(TAG_LENGTH, cekIvBytes);
+    byte[] buffer = new byte[BUFFER_SIZE];
+    Cipher fileCipher = Cipher.getInstance(FILE_CIPHER);
+    fileCipher.init(Cipher.DECRYPT_MODE, fileKey, gcmParameterSpec);
+    if (aad != null) {
+      fileCipher.updateAAD(aad);
+    }
+
+    long totalBytesRead = 0;
+    try (InputStream is = Files.newInputStream(file.toPath(), READ);
+        InputStream cis = new CipherInputStream(is, fileCipher);
+        OutputStream os = Files.newOutputStream(file.toPath(), CREATE)) {
+      int bytesRead;
+      while ((bytesRead = cis.read(buffer)) > -1) {
+        os.write(buffer, 0, bytesRead);
+        totalBytesRead += bytesRead;
+      }
+    }
+
+    // Discard any padding that the encrypted file had
```

Comment:
We shouldn't have padding.

Reply:
So this is something that is not related to padding itself, but to the authentication tag that was stuck at the end of encrypted file.

Reply:
I'm still confused by the comment. Why is there padding?

Reply:
This comment was inaccurate. This is not about padding, this is about auth tag. So when you encrypt a file, auth tag is added at the end of the file. While reading the file, CipherInputStream returns the number of original bytes, without auth tag at the end. If we decrypt file in place, we have to prune unnecessary bytes left from encrypted form.